iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -321,18 +321,138 @@ struct MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm
+///                 %acc_i32 : vector<8x8xi32>,
+///                 [...]
+///
+/// To vector ops reading directly from the %lhs_i8 and %rhs_i8 values
+/// (bypassing the existing arith.extsi) and passing that to a llvm.inline_asm
+/// block implementing the matrix multiplication arithmetic using Aarch64
+/// dot-product instructions (sdot).
+/// It matches the same patterns as MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm
+struct MMT_8x4x8_i8i8i32_Aarch64Dotprod_Intrinsics
+    : public OpRewritePattern<vector::ContractionOp> {
+ public:
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractionOp,
+                                PatternRewriter &rewriter) const override {
+    // It would be nice to sharethe matching logic between this and
+    // MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm, but it defines a bunch of
+    // variables, so doesn't end up being that helpful.
+    if (!isMatrixTimesMatrixTransposedOfGivenShape(contractionOp, 8, 4, 8)) {
+      return failure();
+    }
+
+    Type I8Type = rewriter.getIntegerType(8);
+    Type I32Type = rewriter.getIntegerType(32);
+
+    auto acc = contractionOp.acc();
+    auto lhs = contractionOp.lhs();
+    auto rhs = contractionOp.rhs();
+    if (acc.getType().cast<VectorType>().getElementType() != I32Type) {
+      return failure();
+    }
+
+    Value inLhs = getExtSIInput(I8Type, I32Type, lhs);
+    Value inRhs = getExtSIInput(I8Type, I32Type, rhs);
+
+    if (!inLhs || !inRhs) return failure();
+
+    auto loc = contractionOp.getLoc();
+
+    auto int32x4VType = VectorType::get({4}, I32Type);
+
+    std::array<Value, 16> accChunks;
+    for (int row = 0; row < 8; ++row) {
+      auto accRow =
+          rewriter.create<vector::ExtractOp>(loc, acc, ArrayRef<int64_t>{row});
+      for (int chunk = 0; chunk < 2; ++chunk) {
+        auto accChunk = rewriter.create<vector::ExtractStridedSliceOp>(
+            loc, accRow, ArrayRef<int64_t>{chunk * 4}, ArrayRef<int64_t>{4},
+            ArrayRef<int64_t>{1});
+        assert(accChunk.getType() == int32x4VType);
+        accChunks[row * 2 + chunk] = accChunk;
+      }
+    }
+
+    auto int8x4x4VType = VectorType::get({4, 4}, rewriter.getIntegerType(8));
+    auto extract4x4 = [&](Value in, int rowOffset, int colOffset) {
+      auto chunk = rewriter.create<vector::ExtractStridedSliceOp>(
+          loc, in, ArrayRef<int64_t>{rowOffset, colOffset},
+          ArrayRef<int64_t>{4, 4}, ArrayRef<int64_t>{1, 1});
+      assert(chunk.getType() == int8x4x4VType);
+      return chunk;
+    };
+
+    auto lhs0 = extract4x4(inLhs, 0, 0);
+    auto lhs1 = extract4x4(inLhs, 4, 0);
+    auto rhs0 = extract4x4(inRhs, 0, 0);
+    auto rhs1 = extract4x4(inRhs, 4, 0);
+
+    auto int8Zero4x4 = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(int8x4x4VType));
+    auto sdot = [&](Value acc, Value lhs, Value rhs, int64_t lane) -> Value {
```

Comment:
Please name these "lhs, rhs" parameters of this lambda differently, to avoid confusion as we pass the outer-scope lhs, rhs variables in reverse order here.  For this lambda, it's OK to have meaningless parameter names such as "a" and "b".

Reply:
Done.